xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -430,15 +526,226 @@ private void handleConfigNotFound(StatusException exception) {
+              handleConfigNotFound(error.asException());
+            } else if (!isServing) {
+              listener.onNotServing(error.asException());
+            }
+          }
+        });
+      }
+    }
+
+    private boolean isPermanentError(Status error) {
+      return EnumSet.of(
+              Status.Code.INTERNAL,
+              Status.Code.INVALID_ARGUMENT,
+              Status.Code.FAILED_PRECONDITION,
+              Status.Code.PERMISSION_DENIED,
+              Status.Code.UNAUTHENTICATED)
+              .contains(error.getCode());
+    }
+  }
+
+  @VisibleForTesting
+  final class ConfigApplyingInterceptor implements ServerInterceptor {
+    private final ServerInterceptor noopInterceptor = new ServerInterceptor() {
+      @Override
+      public <ReqT, RespT> Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call,
+          Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+        return next.startCall(call, headers);
+      }
+    };
+
+    @Override
+    public <ReqT, RespT> Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call,
+        Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+      ServerRoutingConfig routingConfig = call.getAttributes().get(ATTR_SERVER_ROUTING_CONFIG);
+      if (routingConfig == null) {
+        call.close(
+                Status.UNAVAILABLE.withDescription("Missing xDS routing config"),
+                new Metadata());
+        return new Listener<ReqT>() {};
+      }
+      VirtualHost virtualHost = RoutingUtils.findVirtualHostForHostName(
+          routingConfig.virtualHosts(), call.getAuthority());
+      if (virtualHost == null) {
+        call.close(
+                Status.UNAVAILABLE.withDescription("Could not find xDS virtual host matching RPC"),
+                new Metadata());
+        return new Listener<ReqT>() {};
+      }
+      Route selectedRoute = null;
+      Map<String, FilterConfig> selectedOverrideConfigs =
+              new HashMap<>(virtualHost.filterConfigOverrides());
+      MethodDescriptor<ReqT, RespT> method = call.getMethodDescriptor();
+      for (Route route : virtualHost.routes()) {
+        if (RoutingUtils.matchRoute(
+            route.routeMatch(), "/" + method.getFullMethodName(), headers, random)) {
+          selectedRoute = route;
+          selectedOverrideConfigs.putAll(route.filterConfigOverrides());
+          break;
+        }
+      }
+      if (selectedRoute == null) {
+        call.close(
+            Status.UNAVAILABLE.withDescription("Could not find xDS route matching RPC"),
+            new Metadata());
+        return new ServerCall.Listener<ReqT>() {};
+      }
+      List<ServerInterceptor> filterInterceptors = new ArrayList<>();
+      for (NamedFilterConfig namedFilterConfig : routingConfig.httpFilterConfigs()) {
+        FilterConfig filterConfig = namedFilterConfig.filterConfig;
+        Filter filter = filterRegistry.get(filterConfig.typeUrl());
+        if (filter instanceof ServerInterceptorBuilder) {
```

Comment:
It's an error if the filter isn't a ServerInterceptorBuilder, but the validation is in a different piece of code. Have an else and close the call with an error (or log, or something) so that we can find any bug?

Reply:
Sure.
[Validation](https://github.com/grpc/grpc-java/blob/1e858921e1d2d3a611d9306833542194880fd715/xds/src/main/java/io/grpc/xds/ClientXdsClient.java#L855) says it would only reach here when `optional`.  Alternatively we may have the validation in xdsClient to be stronger.

Reply:
If it is `optional` we won't reach here as well, because that method never creates the `FilterConfig` and instead returns `null` which discards the entry.